File: packages/shared-contracts/src/inquiry.ts
export const INQUIRY_TYPE_VALUES = ['one_to_one', 'partnership', 'bug_report'] as const;
export const INSTITUTION_TYPE_VALUES = ['school', 'academy', 'other'] as const;
export const ONE_TO_ONE_INQUIRY_CATEGORY_VALUES = [
  'product_usage',
  'account_login',
  'record_upload',
  'other',
] as const;
export const PARTNERSHIP_INQUIRY_CATEGORY_VALUES = ['partnership_request'] as const;
export const BUG_REPORT_INQUIRY_CATEGORY_VALUES = ['bug', 'feature_request'] as const;
export const INQUIRY_CATEGORY_VALUES = [
  ...ONE_TO_ONE_INQUIRY_CATEGORY_VALUES,
  ...PARTNERSHIP_INQUIRY_CATEGORY_VALUES,
  ...BUG_REPORT_INQUIRY_CATEGORY_VALUES,
] as const;

export type InquiryType = (typeof INQUIRY_TYPE_VALUES)[number];
export type InstitutionType = (typeof INSTITUTION_TYPE_VALUES)[number];
export type OneToOneInquiryCategory = (typeof ONE_TO_ONE_INQUIRY_CATEGORY_VALUES)[number];
export type PartnershipInquiryCategory = (typeof PARTNERSHIP_INQUIRY_CATEGORY_VALUES)[number];
export type BugReportInquiryCategory = (typeof BUG_REPORT_INQUIRY_CATEGORY_VALUES)[number];
export type InquiryCategory = (typeof INQUIRY_CATEGORY_VALUES)[number];
export type InquiryMetadataValue = string | number | boolean | null;

export interface InquiryPayload {
  inquiry_type: InquiryType;
  name?: string;
  email: string;
  phone?: string;
  subject?: string;
  message: string;
  inquiry_category?: InquiryCategory;
  institution_name?: string;
  institution_type?: InstitutionType;
  source_path?: string;
  context_location?: string;
  metadata?: Record<string, InquiryMetadataValue>;
}

export interface InquiryResponse {
  id: string;
  inquiry_type: InquiryType;
  status: string;
  delivery_status?: string | null;
  delivery_reason?: string | null;
  delivery_async_job_id?: string | null;
  delivery_retry_needed?: boolean | null;
  created_at: string;
  message: string;
}

export const INQUIRY_ALLOWED_CATEGORIES_BY_TYPE = {
  one_to_one: ONE_TO_ONE_INQUIRY_CATEGORY_VALUES,
  partnership: PARTNERSHIP_INQUIRY_CATEGORY_VALUES,
  bug_report: BUG_REPORT_INQUIRY_CATEGORY_VALUES,
} as const;

export function isInquiryCategoryAllowedForType(
  inquiryType: InquiryType,
  category: InquiryCategory | null | undefined,
): boolean {
  if (!category) {
    return false;
  }
  return INQUIRY_ALLOWED_CATEGORIES_BY_TYPE[inquiryType].includes(
    category as never,
  );
}

File: backend/services/api/src/polio_api/schemas/inquiry.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Any
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

InquiryType = Literal["one_to_one", "partnership", "bug_report"]
InstitutionType = Literal["school", "academy", "other"]
InquiryCategory = Literal[
    "product_usage",
    "account_login",
    "record_upload",
    "partnership_request",
    "bug",
    "feature_request",
    "other",
]

EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class InquiryCreate(BaseModel):
    inquiry_type: InquiryType
    name: str | None = Field(default=None, max_length=120)
    email: str = Field(min_length=5, max_length=255)
    phone: str | None = Field(default=None, max_length=50)
    subject: str | None = Field(default=None, max_length=200)
    message: str = Field(min_length=10, max_length=5000)
    inquiry_category: InquiryCategory | None = None
    institution_name: str | None = Field(default=None, max_length=200)
    institution_type: InstitutionType | None = None
    source_path: str | None = Field(default=None, max_length=255)
    context_location: str | None = Field(default=None, max_length=160)
    metadata: dict[str, Any] | None = None

    @model_validator(mode="after")
    def validate_for_type(self) -> "InquiryCreate":
        self.email = self.email.strip().lower()
        if not EMAIL_PATTERN.match(self.email):
            raise ValueError("A valid email address is required.")

        self.name = _normalize_optional(self.name)
        self.phone = _normalize_optional(self.phone)
        self.subject = _normalize_optional(self.subject)
        self.institution_name = _normalize_optional(self.institution_name)
        self.source_path = _normalize_optional(self.source_path)
        self.context_location = _normalize_optional(self.context_location)
        self.message = self.message.strip()

        if self.inquiry_type == "one_to_one":
            if not self.name:
                raise ValueError("Name is required for one-to-one inquiries.")
            if not self.subject:
                raise ValueError("Subject is required for one-to-one inquiries.")
            if self.inquiry_category not in {"product_usage", "account_login", "record_upload", "other"}:
                raise ValueError("Select a valid inquiry category for one-to-one inquiries.")

        if self.inquiry_type == "partnership":
            if not self.institution_name:
                raise ValueError("Institution name is required for partnership inquiries.")
            if not self.name:
                raise ValueError("Contact name is required for partnership inquiries.")
            if not self.phone:
                raise ValueError("Phone number is required for partnership inquiries.")
            if self.institution_type not in {"school", "academy", "other"}:
                raise ValueError("Select a valid institution type for partnership inquiries.")
            if self.inquiry_category not in {None, "partnership_request"}:
                raise ValueError("Partnership inquiries use the partnership request category.")
            self.inquiry_category = "partnership_request"

        if self.inquiry_type == "bug_report":
            if not self.name:
                raise ValueError("Name or nickname is required for bug reports.")
            if self.inquiry_category not in {"bug", "feature_request"}:
                raise ValueError("Select bug or feature request.")
            if not self.context_location:
                raise ValueError("Context location is required for bug reports.")

        return self


class InquiryCreateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    inquiry_type: InquiryType
    status: str
    delivery_status: str | None = None
    delivery_reason: str | None = None
    delivery_async_job_id: str | None = None
    delivery_retry_needed: bool | None = None
    created_at: datetime
    message: str


def _normalize_optional(value: str | None) -> str | None:
    if value is None:
        return None
    normalized = value.strip()
    return normalized or None
